autoendpoint/src/settings.rs

```diff
@@ -169,10 +199,39 @@ impl Settings {
     }
 }
 
+#[derive(Clone, Debug)]
+pub struct VapidTracker(pub Vec<String>);
+impl VapidTracker {
+    /// Very simple string check to see if the Public Key specified in the Vapid header
+    /// matches the set of trackable keys.
+    pub fn is_trackable(&self, vapid: &VapidHeaderWithKey) -> bool {
+        // ideally, [Settings.with_env_and_config_file()] does the work of pre-populating
+        // the Settings.tracking_vapid_pubs cache, but we can't rely on that.
+        self.0.contains(&vapid.public_key)
+    }
+
+    /// Extract the message Id from the headers (if present), otherwise just make one up.
+    pub fn get_tracking_id(&self, headers: &HeaderMap) -> String {
+        headers
+            .get("X-MessageId")
+            .map(|s| {
+                s.to_str()
+                    .unwrap_or(uuid::Uuid::new_v4().as_simple().to_string().as_str())
+                    .to_string()
+            })
+            .unwrap_or(uuid::Uuid::new_v4().as_simple().to_string())
```

Comment:
These chains can be a bit much but I'd argue utilzing `and_then`/`ok` improves this since we'll only need `new_v4` at the very end (also surprised clippy didn't force `unwrap_or_else` here).
```suggestion
            .and_then(|val| val.to_str().ok())
            .map(|s| s.to_owned())
            .unwrap_or_else(|| uuid::Uuid::new_v4().as_simple().to_string())
```